### PR TITLE
[ROCM] refactoring overload pattern and adapted some ROCM tests

### PR DESCRIPTION
**Pull request by pemeliya**
In this PR I removed Overloaded template used in several places together with std::visit and replaced it with VariantVisitor, and also added GPU architecture checks  for some ROCM tests.

These changes are taken from the original (large) PR: https://github.com/openxla/xla/pull/7782

@xla-rotation: would you please have a look ?

### CODE REVIEW

File: xla/service/gpu/tests/gemm_rewrite_test.cc [hunk 1]
```diff
@@ -984,9 +981,6 @@ ENTRY bf16gemm {
 }
 
 TEST_P(ParameterizedGemmRewriteTest, Int8Gemm) {
-  if (CudaOrRocmCheck(Switch::False, Switch::True)) {
-    GTEST_SKIP() << "DoBlasGemmWithAlgorithm is not yet implemented on ROCm";
```
Comment (dimitar-asenov) [inline xla/service/gpu/tests/gemm_rewrite_test.cc:988]: For this test and the few below it: I don't believe the current PR makes any changes that implement this functionality. Should these checks remain in place for now?

File: xla/stream_executor/device_description.h [hunk 2]
```diff
@@ -220,6 +220,11 @@ class RocmComputeCapability {
   };
 };
 
+// structure supporting C++17 overload pattern:
```
[the remark this answers is not in the document]
Comment (pemeliya) [reply xla/stream_executor/device_description.h:223]: Interesting, I was not aware that this pattern is already available in abseil. Unfortunately, it seems that we pretty old absl version in XLA, in 'functional' there are only any_invocable, bind_front and function_ref available.

If I calculate it correctly, we have the following version currently: https://github.com/abseil/abseil-cpp/tree/lts_2023_08_02

File: xla/service/gpu/tests/gemm_rewrite_test.cc (hunk 1 above)
Comment (pemeliya) [reply xla/service/gpu/tests/gemm_rewrite_test.cc]: yes, sure, I added these checks back again

File: xla/stream_executor/device_description.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (dimitar-asenov) [reply xla/stream_executor/device_description.h:223]: Ah, I see that we update absl whenever there is a new long-term support branch published on Github, which happens once-twice a year. So we won't be able to use the new functionality for now.

Please ignore my suggestion and sorry for the hassle.

**Review by dimitar-asenov — APPROVED** [review]

**Pull request closed without merging**